pkg/controller/workspace/runtime/runtime_annotation.go

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package runtime
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/che-incubator/che-workspace-operator/pkg/apis/workspace/v1alpha1"
+)
+
+func ConstructRuntimeAnnotation(components []v1alpha1.ComponentDescription, endpoints map[string][]v1alpha1.ExposedEndpoint) (string, error) {
+	defaultEnv := "default"
+
+	machines := getMachinesAnnotation(components, endpoints)
+	commands := getWorkspaceCommands(components)
+
+	runtime := v1alpha1.CheWorkspaceRuntime{
+		ActiveEnv: defaultEnv,
+		Commands:  commands,
+		Machines:  machines,
+	}
+
+	runtimeJSON, err := json.Marshal(runtime)
+	if err != nil {
+		return "", err
+	}
+	return string(runtimeJSON), nil
+}
+
+func getMachinesAnnotation(components []v1alpha1.ComponentDescription, endpoints map[string][]v1alpha1.ExposedEndpoint) map[string]v1alpha1.CheWorkspaceMachine {
+	machines := map[string]v1alpha1.CheWorkspaceMachine{}
+
+	for _, component := range components {
+		for containerName, container := range component.ComponentMetadata.Containers {
+			servers := map[string]v1alpha1.CheWorkspaceServer{}
+			// TODO: This is likely not a good choice for matching, since it'll fail if container name does not match an endpoint key
+			for _, endpoint := range endpoints[containerName] {
+				protocol := endpoint.Attributes[v1alpha1.PROTOCOL_ENDPOINT_ATTRIBUTE]
+
+				servers[endpoint.Name] = v1alpha1.CheWorkspaceServer{
+					Attributes: endpoint.Attributes,
+					Status:     v1alpha1.RunningServerStatus, // TODO: This is just set so the circles are green
+					URL:        fmt.Sprintf("%s://%s", protocol, endpoint.Url),
```

Comment:
It leads to url with double protocol like `"http://https://workspace383a324431b54d9f-theia-proxy-4400.apps-crc.testing`

Something like https://github.com/amisevsk/che-workspace-crd-operator/commit/84dd6135e7f777033ea1faec8d73a6dc585f9ade could be done to solve this issue.

Reply:
This is an artifact of me implementing Josh's request for protocol in status -- I forgot to remove it here. It doesn't work properly anyways, since protocol is *always* `http` even for https routes. We should have protocol included in URL everywhere now, so checking is unnecessary.

Reply:
Ah except when it doesn't:
```
ws://http://workspace82ba65b9c2b84595-che-machine-exec-4444.apps-crc.testing
```

Reply:
So, now we get https URL even when protocol is ws, right?